Firestore/core/test/firebase/firestore/local/counting_query_engine.h

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_TEST_FIREBASE_FIRESTORE_LOCAL_COUNTING_QUERY_ENGINE_H_
+#define FIRESTORE_CORE_TEST_FIREBASE_FIRESTORE_LOCAL_COUNTING_QUERY_ENGINE_H_
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/core/query.h"
+#include "Firestore/core/src/firebase/firestore/local/mutation_queue.h"
+#include "Firestore/core/src/firebase/firestore/local/query_engine.h"
+#include "Firestore/core/src/firebase/firestore/local/remote_document_cache.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/document_map.h"
+#include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+class LocalDocumentsView;
+class WrappedMutationQueue;
+class WrappedRemoteDocumentCache;
+
+/**
+ * A test-only QueryEngine that forwards all API calls and exposes the number of
+ * documents and mutations read.
+ */
+class CountingQueryEngine : public QueryEngine {
+ public:
+  CountingQueryEngine(QueryEngine* query_engine, bool index_free)
+      : query_engine_(query_engine), index_free_(index_free) {
+  }
+
+  void ResetCounts();
+
+  void SetLocalDocumentsView(LocalDocumentsView* local_document) override;
+
+  model::DocumentMap GetDocumentsMatchingQuery(
+      const core::Query& query,
+      const model::SnapshotVersion& last_limbo_free_snapshot_version,
+      const model::DocumentKeySet& remote_keys) override;
+
+  /**
+   * Returns whether the backing query engine is optimized to perform key-based
+   * lookups.
+   */
+  // TODO(mrschmidt): Come up with a name that describes the behavior change
+  bool is_index_free() {
```

Comment:
This does not exist on Android, where I am determining the mode using an instanceof check.

Reply:
Elsewhere in the C++ SDK where we need instanceof checks we have an `enum class` of the various kinds we care about and then a `type()` accessor that returns that enum value. In this case we would have a `QueryEngine::Type` enum.

Reply:
I updated the PR to follow the `type()` pattern and added a `QueryEngine::Type` enum.